src/libfetchers/tests/attrs.cc

```diff
@@ -0,0 +1,284 @@
+    }
+
+    TEST(getIntAttr, throwsWhenValueIsNotAnInt) {
+        Attrs emptyAttrs = {
+          { "int", "42" }
+        };
+
+        ASSERT_THROW(getIntAttr(emptyAttrs, "int"), Error);
+    }
+
+    /* ----------------------------------------------------------------------------
+     * maybeGetBoolAttr
+     * --------------------------------------------------------------------------*/
+
+    TEST(maybeGetBoolAttr, throwsOnEmptyAttr) {
+        Attrs emptyAttrs = { };
+
+        auto res = maybeGetBoolAttr(emptyAttrs, "bool");
+        ASSERT_FALSE(res.has_value());
+
+    }
+
+    TEST(maybeGetBoolAttr, getsBoolFromAttr) {
+        Attrs emptyAttrs = {
+          { "bool", true }
+        };
+
+        auto res = maybeGetBoolAttr(emptyAttrs, "bool");
+        ASSERT_TRUE(res.has_value());
+        ASSERT_EQ(res.value(), true);
+    }
+
+    TEST(maybeGetBoolAttr, throwsWhenValueIsNotABool) {
+        Attrs emptyAttrs = {
+          { "bool", "42" }
+        };
+
+        ASSERT_THROW(maybeGetBoolAttr(emptyAttrs, "bool"), Error);
+    }
+
+    /* ----------------------------------------------------------------------------
+     * getBoolAttr
+     * --------------------------------------------------------------------------*/
+
+    TEST(getBoolAttr, throwsOnEmptyAttr) {
+        Attrs emptyAttrs = { };
+
+        ASSERT_THROW(getBoolAttr(emptyAttrs, "bool"), Error);
+    }
+
+    TEST(getBoolAttr, getsBoolFromAttr) {
+        Attrs emptyAttrs = {
+          { "bool", true }
+        };
+
+        ASSERT_EQ(getBoolAttr(emptyAttrs, "bool"), true);
+    }
+
+    TEST(getBoolAttr, throwsWhenValueIsNotABool) {
+        Attrs emptyAttrs = {
+          { "bool", "42" }
+        };
+
+        ASSERT_THROW(getBoolAttr(emptyAttrs, "bool"), Error);
+    }
+
+    TEST(getBoolAttr, getsNonZeroIntsCastToTrue) {
+        Attrs emptyAttrs = {
+          { "bool", 42 }
+        };
+
+        ASSERT_EQ(getBoolAttr(emptyAttrs, "bool"), true);
+    }
+
+    TEST(getBoolAttr, getsZeroIntsCastToFalse) {
+        Attrs emptyAttrs = {
+          { "bool", 0 }
+        };
+
+        ASSERT_EQ(getBoolAttr(emptyAttrs, "bool"), false);
+    }
+
+    /* ----------------------------------------------------------------------------
+     * attrsToQuery
+     * --------------------------------------------------------------------------*/
+
+    TEST(attrsToQuery, returnsEmptyMapOnEmptyAttrs) {
+        Attrs emptyAttrs = { };
+        std::map<std::string, std::string> emptyMap = { };
+
+        ASSERT_EQ(attrsToQuery(emptyAttrs), emptyMap);
+    }
+
+    TEST(attrsToQuery, stringifiesAttr) {
+
+        Explicit<bool> b = { true };
+        Attrs attrs = {
+          { "num", 42 },
+          { "string", "this is a string" },
+          { "bool", b }
+        };
+
+        std::map<std::string, std::string> expected = {
+            { "num", "42" },
+            { "string", "this is a string" },
+            { "bool", "1" }
```

Comment:
I do wonder why the boolean is being converted to `"1"` or `"0"` here instead of `true` or `false`..

Reply:
IIRC this is fixed on the flakes branch, which rewrites most of libfetchers. Please rebase on that branch. BTW, that's an example of the cost of unit tests: you end up writing a lot of code that has to be discarded when the API changes.

Reply:
> BTW, that's an example of the cost of unit tests: you end up writing a lot of code that has to be discarded when the API changes.

I think Nix might be the only current project that I know where unit tests are regarded a controversial topic. 

The way everyone else seems to look at it is that unit tests are a net win because they provide confidence when making changes and thus effectively making API changes easier and safer.